src/System.Private.CoreLib/src/System/Threading/ThreadPool.cs

```diff
@@ -533,52 +1328,71 @@ internal static bool Dispatch()
+            // we need to return to the VM.
+            //
+            int quantumStartTime = Environment.TickCount;
+
             // Has the desire for logging changed since the last time we entered?
-            outerWorkQueue.loggingEnabled = FrameworkEventSource.Log.IsEnabled(EventLevel.Verbose, FrameworkEventSource.Keywords.ThreadPool | FrameworkEventSource.Keywords.ThreadTransfer);
+            var enabled = FrameworkEventSource.Log.IsEnabled(EventLevel.Verbose, FrameworkEventSource.Keywords.ThreadPool | FrameworkEventSource.Keywords.ThreadTransfer);
+            if (outerWorkQueue.loggingEnabled != enabled)
+            {
+                // writing shared state.
+                outerWorkQueue.loggingEnabled = enabled;
+            }
 
             //
             // Assume that we're going to need another thread if this one returns to the VM.  We'll set this to 
             // false later, but only if we're absolutely certain that the queue is empty.
             //
             bool needAnotherThread = true;
             object outerWorkItem = null;
-            try
-            {
-                //
-                // Set up our thread-local data
-                //
-                // Use operate on workQueue local to try block so it can be enregistered 
-                ThreadPoolWorkQueue workQueue = outerWorkQueue;
-                ThreadPoolWorkQueueThreadLocals tl = workQueue.GetOrCreateThreadLocals();
-                Thread currentThread = tl.currentThread;
 
-                // Start on clean ExecutionContext and SynchronizationContext
-                currentThread.ExecutionContext = null;
-                currentThread.SynchronizationContext = null;
+            // On first successfull Dequeue we ask for another thread.
+            // If the new thread that comes is sucessful, it will ask for more and so on...
+            //
+            // We will repeat the expansion request after some number of sucesses. As long as we find work to do.
+            // If a thread finds no work due to shortage or contentions, it will exit to VM and park there.
+            //
+            int dequeueCycles = 0;
+
+            //
+            // Use operate on workQueue local to try block so it can be enregistered 
+            ThreadPoolWorkQueue workQueue = outerWorkQueue;
+            Thread currentThread = Thread.CurrentThread;
+
+            // Start on clean ExecutionContext and SynchronizationContext
+            currentThread.ExecutionContext = null;
+            currentThread.SynchronizationContext = null;
 
+            try
+            {
+                // refresh proc ID for the new quantum.
+                // also, since we are coming from the VM, we could have a context switch when going through the semaphore.
+                var localQueueIndex = workQueue.GetLocalQueueIndex(Threading.Thread.RefreshCurrentProcessorId());
```

Comment:
> My reasoning was that when we come from VM we may consider that long enough time passed since we last queried and the counter on the cache no longer reflects "oldness" of the data.

Yes that assertion is not true in bursty cases. There, threads frequently leave `Dispatch` and reenter `Dispatch` soon afterwards.

I think the sustained case still needs a `Refresh` as I described above, otherwise the ID could become stale and remain that way for a long time.

Reply:
> In sustained case (when we are looping) we would be checking the value frequently

As far as I can tell the queue index based on current processor number is determined before the loop, so I don't see how in a sustained case the value would be checked frequently. I believe it should be done outside the loop, but without a refresh on quantum expiration, the frequency at which `GetCurrentProcessorId()` would be called would be once per quantum, which is quite long when it takes 5000 or such calls to actually refresh.

Reply:
Turns out that moving core ID refresh to the end of the `Dispatch` has very small effect on bursty scenarios (0-5% improvements on few scenarios, but mostly no change). At the same time the change causes big regressions in throughput. Ben's benchmark shows 10% or more regression across the whole local queue matrix.

There could be multiple explanations for this - like going through VM could be the most likely place where threads switch cores (if they ever switch) so benefits from having the refresh right after transitioning from VM into managed `Dispatch` outweighs expenses.

Anyways, it would be interesting to find out exactly what the reason(s) are, but so far the measurements indicate that entering `Dispatch` is a good place for refreshing core ID. 
I'll keep it there.

Reply:
0-5% doesn't seem like a small amount to me. Perhaps there are other ways to fix it. I would imagine that getting the proc ID from the native side and passing it into `Dispatch` would be a lot more efficient. It could perhaps be further improved by refreshing the proc ID from the native side only when the thread actually wakes up from a wait.

Reply:
Yes, I thought the same way too. The most optimal approach could be if core ID is refreshed after a thread resumes running. The issue is that the ID is stored on the managed side while the worker semaphore is on the native side, which I am trying not to change.  

Another possibility is to pass a parameter to `Dispatch` indicating whether the thread was resumed. That could be a smaller change.

Reply:
Other thoughts regarding implementation of `RefreshCurrentProcessorId()`:
- It would make sense for the check for negative value and using the managed thread ID instead to be taken care of by the native side as part of `GetCurrentProcessorNumber()`
- I'm not sure what the point of this is:
  ```c#
            // Add offset to make it clear that it is not guaranteed to be 0-based processor number
            currentProcessorId += 100;
  ```
  But if there's no good reason for it I would just remove it. It's probably not that important to change it anyway.
- It would also be fairly easy to generate the same type of ID from the native side

Another consideration is that CoreRT's Unix thread pool implementation, which is being ported to CoreCLR (to the shared partition for use in Mono) does not use a quantum in `Dispatch`, that is, it stays in `Dispatch` as long as there are work items. With that implementation, the managed side would need to update the proc ID every so often despite not having to return from `Dispatch`, so it may be beneficial for the native side to generate the same type of ID as `RefreshCurrentProcessorId()`.